Assets/MixedRealityToolkit/Services/BaseEventSystem.cs

```diff
@@ -25,44 +39,226 @@ public abstract class BaseEventSystem : BaseService, IMixedRealityEventSystem
         public virtual void HandleEvent<T>(BaseEventData eventData, ExecuteEvents.EventFunction<T> eventHandler) where T : IEventSystemHandler
         {
             Debug.Assert(!eventData.used);
+
             eventExecutionDepth++;
 
+            // First handle events via old API part, via GameObjects.
+            // This sends event to every component of matching type on this object, 
+            // INCLUDING the ones that don't expect it.
             for (int i = EventListeners.Count - 1; i >= 0; i--)
             {
                 ExecuteEvents.Execute(EventListeners[i], eventData, eventHandler);
             }
 
+            // Send events to all handlers registered via RegisterHandler API.
+            HashSet<IEventSystemHandler> handlers;
+            if (EventHandlersByType.TryGetValue(typeof(T), out handlers))
+            {
+                foreach (IEventSystemHandler handler in handlers)
+                {
+                    // If handler is a unity component, need to make sure, that it didn't receive an event
+                    // when we were sending it to Game objects above.
+                    // Do that by checking if its parent is registered in EventListeners.
+                    var componentHandler = handler as Component;
```

Comment:
I don't have any data here to back this up, but it feels weird having to do a null check + hashset lookup within each component call.

I would almost rather have us be explicit in being like you can either register for one or the other, so that you have to explicitly move to the new one or stay on the old one.

The only value I can see in having both is if you might use it during migration - but even then it's questionable - if you've registered for both, your "new method" registration is basically useless, and you can't validate it works until you unregister from the old, and if we support both at the same time and then kill off the old in a future release, the customer would have never been forced to evaluate the new method entirely (i.e. will suddenly stop getting messages for another one)

Reply:
The problem comes from having 2 different components on the same object. If one of them uses old API, and second uses new API, second will be triggered during `ExecuteEvents` call. We need this check to avoid second invoke of the callback.

Reply:
right right - I think what I'm saying is, maybe we could just change the Register functions to throw if you tried to register via the old way AND the new way - basically, let's just not support the scenario and error out as early as possible (and do the checking in a place that happens not in every tick), and that we can also reduce some of the complexity here (i.e. if we prevent old + new registration, there's no need to do these checks here)

Reply:
This might cause weird behavior, when users just add a new component to an object in existing scene and things stop working because there's an API conflict.
I'll see if I can cache that.

Reply:
Actually I'll leave it as is for now, because overhead of `ExecuteEvents.Execute` is at least comparable if not much more.

Reply:
I think we should consider an error to register a game object and a handler in one of its components, so I wouldn't add code to make it silently work. Instead I'd check as much as we can on registration, as Will was suggesting.

Reply:
Let's do this as a part of #4847. Right now the majority of Mrtk components still uses old API (I can't enable Obsolete warning because of that). Once those are cleaned up, we can mark methods as obsolete and add stricter validation on in `Register` / `Unregister` implementation, which will only remain in client code.

Reply:
sounds good to me!